### DISCUSSION TITLE
Data persistency on shared hosting

### DISCUSSION THREAD

**Discussion by johnd0e** [umputun/tg-spam#145] [Ideas]
Consider the following use case: the only aspect under user control is the Dockerfile.
To modify an option or update the `tg-spam` application, a new (edited) Dockerfile must be redeployed, but this process does not preserve any existing data.

Here are a couple of ideas that come to mind:

1. **Synchronization Job**: We could implement a synchronization job, for example, using `rclone`. However, we would need a mechanism to run it periodically. Perhaps `tg-spam` could handle this task.

2. **Cloud Database**: Another option would be to use a cloud database instead of relying on text files. There are some free options available, such as Cloudflare R2, which is S3-compatible.

Do you have any suggestions for this scenario?

**Comment by johnd0e** [comment]
> Synchronization Job

I found some suggestions on [Stack Overflow](https://superuser.com/questions/181517/how-to-execute-a-command-whenever-a-file-changes), and [inotifywait](https://github.com/inotify-tools/inotify-tools/wiki) seems to be the most promising option.

> Cloud Database

Alternatively, we could use SQLite with cloud replication through [Litestream](https://litestream.io/).
I discovered this suggestion [here](https://mtlynch.io/litestream/).

**Reply by umputun** [reply]
Switching all the persistency to the database layer will address the issue you are describing and, in addition, will simplify the code. However, it will also make direct editing of sample and blacklist files impossible unless we introduce some automatic conversion from files to database records, which will add some complexity and a need to sync two sources inside the service.

I'm not sure yet if this trade-off is worth it.

**Reply by umputun** [reply]
And, because in the hosting environment you are describing, it has no access to a shared volume, the initial preparation and loading of lists is another problem. This will probably need new API endpoints (i.e., `POST /spamfile` or something like this) that get the whole file in the request body and the corresponding change in the UI.

**Reply by johnd0e** [reply]
> unless we introduce some automatic conversion from files to database records, which will add some complexity

Given that the format of our text files is quite simple, transforming text lines into SQL statements (and vice versa) should be relatively straightforward.
Alternatively, we can also consider using the [sqlean-fileio](https://dev.to/nalgeon/reading-and-writing-files-in-sqlite-32ia) extension.

I believe there should be a single source of truth: the database. The text files could then serve solely for data interchange.

> I'm not sure yet if this trade-off is worth it.

Well, here is only an idea, you decide.

> the initial preparation and loading of lists is another problem.

In the scenario with SQLite and Litestream, it can be resolved simply by adding a command to the Dockerfile.
But I agree that the new API endpoint can be useful for other use cases.

**Reply by umputun** [reply]
> I believe there should be a single source of truth: the database. The text files could then serve solely for data interchange.

I agree, but we still have to support an automatic conversion of all the existing lists. In addition (this one is rather nice to have), we need the ability to upload new data in the same text format if the user wishes to update lists in batch mode with some strategy for confilcts resolution . We probably need to do it the opposite way as well, which is another part that needs some development.

**Reply by johnd0e** [reply]
> an automatic conversion of all the existing lists.

This process could be permanent: it would monitor for new files, import them, and delete the processed ones.

>  if the user wishes to update lists in batch mode with some strategy for confilcts resolution

I propose the following simple strategy:

- The new file replaces the entire database table.
- If we decide to support some form of updating, we could use files with the suffix `_update`.
- Since we don't have IDs in the text, the only viable conflict resolution method would be to ensure no duplicates.

> We probably need to do it the opposite way as well

Dump db tables and return them as files on web-request? Or something different?

> ...which is another part that needs some development.

Indeed, it appears to be a substantial amount of work.

**Comment by johnd0e** [comment]
The problem is solved, as tg-spam from master switched to sqlite.
Now I can achieve persistency with `litestream`:

```dockerfile
FROM ghcr.io/umputun/tg-spam:master

#ENV LITESTREAM_ACCESS_KEY_ID=01234567890
#ENV LITESTREAM_SECRET_ACCESS_KEY=qWeRtYuIoP
#ENV REPLICA_URL=s3://tg-spam.s3.us-east-000.backblazeb2.com/db

COPY <<'EOF' entrypoint.sh
#!/bin/sh
FILES_DYNAMIC="${FILES_DYNAMIC:-data}"
DB_PATH="$FILES_DYNAMIC/tg-spam.db"
./litestream restore -o $DB_PATH $REPLICA_URL
if [ ! -f "$DB_PATH" ]; then
  echo "tg-spam.db not found, copying preset files to $FILES_DYNAMIC"
  cp -r preset/* "$FILES_DYNAMIC"
fi
CMD="./tg-spam --first-messages-count=2 --openai.retry-count=3 --dbg \$@"
exec ./litestream replicate -exec "$CMD" $DB_PATH $REPLICA_URL
EOF

USER root
RUN <<EOF
chmod +x entrypoint.sh
mkdir logs
apk add --no-cache jq
wget -qO- "$(wget -qO- "https://api.github.com/repos/benbjohnson/litestream/releases/latest" \
             | jq -r '.assets[] | select(.name | contains("-linux-amd64.tar.gz")) | .browser_download_url')" \
     | tar xzf -
chmod +x litestream
EOF
USER app
```
